subsys/lorawan/nvm/lorawan_nvm_settings.c

```diff
@@ -0,0 +1,179 @@
+	NVM_SETTING_DESCR(LORAMAC_NVM_NOTIFY_FLAG_MAC_GROUP2, MacGroup2),
+	NVM_SETTING_DESCR(LORAMAC_NVM_NOTIFY_FLAG_SECURE_ELEMENT, SecureElement),
+	NVM_SETTING_DESCR(LORAMAC_NVM_NOTIFY_FLAG_REGION_GROUP1, RegionGroup1),
+	NVM_SETTING_DESCR(LORAMAC_NVM_NOTIFY_FLAG_REGION_GROUP2, RegionGroup2),
+	NVM_SETTING_DESCR(LORAMAC_NVM_NOTIFY_FLAG_CLASS_B, ClassB),
+};
+
+static void lorawan_nvm_save_settings(uint16_t nvm_notify_flag)
+{
+	MibRequestConfirm_t mib_req;
+
+	LOG_DBG("Saving LoRaWAN settings");
+
+	/* Retrieve the actual context */
+	mib_req.Type = MIB_NVM_CTXS;
+	if (LoRaMacMibGetRequestConfirm(&mib_req) != LORAMAC_STATUS_OK) {
+		LOG_ERR("Could not get NVM context");
+		return;
+	}
+
+	LoRaMacNvmData_t *nvm = mib_req.Param.Contexts;
+
+	LOG_DBG("Crypto version: %"PRIu32", DevNonce: %d, JoinNonce: %"PRIu32,
+		mib_req.Param.Contexts->Crypto.LrWanVersion.Value,
+		mib_req.Param.Contexts->Crypto.DevNonce,
+		mib_req.Param.Contexts->Crypto.JoinNonce);
+
+	for (uint32_t i = 0; i < ARRAY_SIZE(nvm_setting_descriptors); i++) {
+		const struct lorawan_nvm_setting_descr *descr =
+			&nvm_setting_descriptors[i];
+
+		if ((nvm_notify_flag & descr->flag) == descr->flag) {
+			LOG_DBG("Saving configuration %s", descr->setting_name);
+			int err = settings_save_one(descr->setting_name,
+						(char *)nvm + descr->offset,
+						descr->size);
+			if (err) {
+				LOG_ERR("Could not save settings %s, error %d",
+					descr->name, err);
+			}
+		}
+	}
+
+	settings_save();
+}
+
+void lorawan_nvm_data_mgmt_event(uint16_t flags)
+{
+	if (flags != LORAMAC_NVM_NOTIFY_FLAG_NONE) {
+		lorawan_nvm_save_settings(flags);
+	}
+}
+
+static int load_setting(void *tgt, size_t tgt_size,
+			const char *key, size_t len,
+			settings_read_cb read_cb, void *cb_arg)
+{
+	if (len != tgt_size) {
+		LOG_ERR("Can't load '%s' state, size mismatch.",
+			log_strdup(key));
+		return -EINVAL;
+	}
+
+	if (!tgt) {
+		LOG_ERR("Can't load '%s' state, no target.",
+			log_strdup(key));
+		return -EINVAL;
+	}
+
+	if (read_cb(cb_arg, tgt, len) != len) {
+		LOG_ERR("Can't load '%s' state, short read.",
+			log_strdup(key));
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int on_setting_loaded(const char *key, size_t len,
+			   settings_read_cb read_cb,
+			   void *cb_arg, void *param)
+{
+	int err;
+	LoRaMacNvmData_t *nvm = param;
+
+	LOG_DBG("Key: %s", log_strdup(key));
+
+	for (uint32_t i = 0; i < ARRAY_SIZE(nvm_setting_descriptors); i++) {
+		const struct lorawan_nvm_setting_descr *descr =
+			&nvm_setting_descriptors[i];
+
+		if (strcmp(descr->name, key) == 0) {
+			err = load_setting((char *)nvm + descr->offset,
+				descr->size, key, len, read_cb, cb_arg);
+			if (err) {
+				LOG_ERR("Could not read setting %s", descr->name);
+			}
+			return err;
+		}
+	}
+
+	LOG_WRN("Unknown LoRaWAN setting: %s", log_strdup(key));
+	return 0;
+}
+
+int lorawan_nvm_data_restore(void)
+{
+	int err;
+	LoRaMacStatus_t status;
+	MibRequestConfirm_t mib_req;
+
```

Comment:
Not a big deal, but adding a print would help during debugging:

LOG_DBG("Restoring LoRaWAN settings");